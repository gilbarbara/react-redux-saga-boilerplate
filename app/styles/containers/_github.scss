.app__github {
  &__search {
    margin: ($grid-gutter-width / 2) 0;
    text-align: center;

    @include lg {
      margin-bottom: $grid-gutter-width;
    }
  }

  &__userlist {
    margin: ($grid-gutter-width / 2) 0;
  }

  &__grid {
    display: grid;
    grid-auto-flow: row;
    grid-gap: $grid-gutter-width / 3;
    grid-template-columns: 100%;
    list-style: none;
    margin: 0;
    padding: 0;

    @include sm {
      grid-template-columns: repeat(2, 1fr);
    }

    @include lg {
      grid-gap: $grid-gutter-width / 2;
      grid-template-columns: repeat(3, 1fr);
    }

    @include xl {
      grid-gap: $grid-gutter-width;
      grid-template-columns: repeat(4, 1fr);
    }

    > li {
      display: flex;
    }
  }

  &__item {
    align-items: center;
    border: solid 0.1rem $app-color;
    border-radius: 0.4rem;
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    overflow: hidden;
    padding: $grid-gutter-width / 3;
    text-align: center;
    width: 100%;

    @include md {
      padding: $grid-gutter-width / 2;
    }

    @include lg {
      padding: $grid-gutter-width;
    }

    img {
      height: 8rem;
      margin-bottom: $grid-gutter-width / 2;
    }

    > div {
      flex: 1;
      margin-left: $grid-gutter-width / 2;
    }

    h5 {
      line-height: 1;
      margin-bottom: $grid-gutter-width / 3;

      @include ix-only {
        font-size: $font-size-base;
      }

      a {
        display: block;
      }

      small {
        color: $gray-600;
      }
    }
  }

  .btn-primary {
    pointer-events: none;
  }
}
