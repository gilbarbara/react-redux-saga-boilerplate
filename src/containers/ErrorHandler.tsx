import React from 'react';
import { Container, Heading, Image } from 'styled-minimal';

type Props = {
  children?: any;
  onError?: (error: Error, componentStack: string) => void;
};

type ErrorInfo = {
  componentStack: string;
};

type State = {
  error: Error | null;
};

export default class ErrorHandler extends React.Component<Props, State> {
  public state: State = {
    error: null,
  };

  componentDidCatch(error: Error, info: ErrorInfo) {
    const { onError } = this.props;

    /* istanbul ignore else */
    if (typeof onError === 'function') {
      try {
        onError.call(this, error, info?.componentStack);
      } catch {
        // ignore
      }
    }

    this.setState({ error });
  }

  render() {
    const { children } = this.props;
    const { error } = this.state;

    if (error === null) {
      return children;
    }

    const message = error.toString();

    return (
      <Container bg="#000" cl="#fff" fullScreen textAlign="center" data-testid="ErrorHandler">
        <Image src={`${process.env.PUBLIC_URL}/media/images/error.gif`} alt="Error" width={128} />
        <Heading as="h3" my={3}>
          {message}
        </Heading>
      </Container>
    );
  }
}
