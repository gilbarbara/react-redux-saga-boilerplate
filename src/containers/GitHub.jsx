import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import styled from 'styled-components';
import treeChanges from 'tree-changes';
import { appColor } from 'modules/theme';

import { getRepos, showAlert, switchMenu } from 'actions/index';
import { STATUS } from 'constants/index';

import {
  Box,
  ButtonGroup,
  Button,
  Flex,
  Heading,
  Link,
  Image,
  Paragraph,
  theme,
  utils,
} from 'styled-minimal';
import Loader from 'components/Loader';

const { responsive, spacer } = utils;
const { grays } = theme;

const GitHubWrapper = styled.div``;

const GitHubGrid = styled.ul`
  display: grid;
  grid-auto-flow: row;
  grid-gap: ${spacer(2)};
  grid-template-columns: 100%;
  list-style: none;
  margin: ${spacer(4)} auto 0;
  padding: 0;
  ${/* istanbul ignore next */p => responsive({
    ix: `
      grid-gap: ${spacer(3)(p)};
      width: 90%;
    `,
    md: `
      grid-template-columns: repeat(2, 1fr);
      width: 100%;
    `,
    lg: `
      grid-template-columns: repeat(3, 1fr);
    `,
    xl: `
      grid-gap: ${spacer(4)(p)};
      grid-template-columns: repeat(4, 1fr);
    `,
  })};

  > li {
    display: flex;
  }
`;

const Item = styled(Box)`
  align-items: center;
  border: solid 0.1rem ${appColor};
  border-radius: 0.4rem;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  overflow: hidden;
  padding: ${spacer(3)};
  text-align: center;
  width: 100%;
  ${/* istanbul ignore next */p => responsive({
    md: `
      padding: ${spacer(3)(p)};
    `,
    lg: `
      padding: ${spacer(4)(p)};
    `,
  })};

  > a {
    margin-bottom: ${spacer(2)};
  }

  img {
    height: 8rem;
  }
`;

const ItemHeader = styled.div`
  margin-bottom: ${spacer(3)};
  
  a {
    display: block;
  }

  small {
    color: ${grays.gray60};
  }
`;

export class GitHub extends React.Component {
  state = {
    query: 'react',
  };

  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    github: PropTypes.object.isRequired,
  };

  componentDidMount() {
    const { query } = this.state;
    const { dispatch } = this.props;

    dispatch(getRepos(query));
  }

  componentWillReceiveProps(nextProps) {
    const { dispatch } = this.props;
    const { changedTo } = treeChanges(this.props, nextProps);

    if (changedTo('github.repos.status', STATUS.ERROR)) {
      dispatch(showAlert(nextProps.github.repos.message, { variant: 'danger' }));
    }
  }

  handleClick = (e) => {
    const { query } = e.currentTarget.dataset;
    const { dispatch } = this.props;

    this.setState({
      query,
    });

    dispatch(switchMenu(query));
  };

  render() {
    const { query } = this.state;
    const { github } = this.props;
    const data = github.repos.data[query] || [];
    let output;

    if (github.repos.status === STATUS.READY) {
      if (data.length) {
        output = (
          <GitHubGrid data-type={query}>
            {github.repos.data[query]
              .map(d => (
                <li key={d.id}>
                  <Item>
                    <Link href={d.html_url}>
                      <Image src={d.owner.avatar_url} alt={d.owner.login} />
                    </Link>
                    <ItemHeader>
                      <Link href={d.html_url}>
                        <Heading as="h5" lineHeight={1}>{d.name}</Heading>
                        <small>{d.owner.login}</small>
                      </Link>
                    </ItemHeader>
                    <Paragraph>{d.description}</Paragraph>
                  </Item>
                </li>
              ))}
          </GitHubGrid>
        );
      }
      else {
        output = <h3>Nothing found</h3>;
      }
    }
    else {
      output = <Loader />;
    }

    return (
      <GitHubWrapper key="GitHub">
        <Flex justifyContent="center">
          <ButtonGroup role="group" aria-label="GitHub Selector">
            <Button
              animate={query === 'react' && github.repos.status === 'running'}
              outline={query !== 'react'}
              size="lg"
              data-query="react"
              onClick={this.handleClick}
            >
              React
            </Button>
            <Button
              animate={query === 'redux' && github.repos.status === 'running'}
              outline={query !== 'redux'}
              size="lg"
              data-query="redux"
              onClick={this.handleClick}
            >
              Redux
            </Button>
          </ButtonGroup>
        </Flex>
        {output}
      </GitHubWrapper>
    );
  }
}

/* istanbul ignore next */
function mapStateToProps(state) {
  return { github: state.github };
}

export default connect(mapStateToProps)(GitHub);
