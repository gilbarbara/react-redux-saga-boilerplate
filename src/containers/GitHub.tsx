import React, { useCallback } from 'react';
import { useDispatch } from 'react-redux';
import { useMount, usePrevious, useUpdateEffect } from 'react-use';
import styled from 'styled-components';
import {
  Button,
  ButtonGroup,
  Flex,
  Grid,
  Heading,
  Image,
  Link,
  Paragraph,
  responsive,
} from 'styled-minimal';

import { useShallowEqualSelector } from 'modules/hooks';
import theme, { appColor, spacer } from 'modules/theme';

import { getRepos, showAlert } from 'actions';
import { STATUS } from 'literals';

import Loader from 'components/Loader';

const Item = styled(Link)`
  align-items: center;
  border: solid 0.1rem ${appColor};
  border-radius: 0.4rem;
  display: flex;
  flex-direction: column;
  overflow: hidden;
  padding: ${spacer(3)};
  text-align: center;
  width: 100%;
  /* stylelint-disable */
  ${responsive({
    md: {
      padding: spacer(3),
    },
    lg: {
      padding: spacer(4),
    },
  })};
  /* stylelint-enable */

  p {
    color: #000;
  }

  img {
    height: 8rem;
    margin-bottom: ${spacer(2)};
  }
`;

const ItemHeader = styled.div`
  margin-bottom: ${spacer(3)};

  small {
    color: ${theme.colors.gray60};
  }
`;

function GitHub() {
  const dispatch = useDispatch();
  const { data, message, query, status } = useShallowEqualSelector(({ github }) => ({
    data: github.topics[github.query]?.data || [],
    message: github.topics[github.query]?.message || '',
    query: github.query,
    status: github.topics[github.query]?.status || STATUS.IDLE,
  }));
  const previousStatus = usePrevious(status);

  useMount(() => {
    dispatch(getRepos(query || 'react'));
  });

  useUpdateEffect(() => {
    if (previousStatus !== status && status === STATUS.ERROR) {
      dispatch(showAlert(message, { variant: 'danger' }));
    }
  }, [message, previousStatus, status]);

  const handleClick = useCallback(
    (e: React.MouseEvent<HTMLButtonElement>) => {
      const { topic = '' } = e.currentTarget.dataset;

      dispatch(getRepos(topic));
    },
    [dispatch],
  );

  const isRunning = status === STATUS.RUNNING;
  let output;

  if (status === STATUS.SUCCESS) {
    if (data.length) {
      output = (
        <Grid
          gridGap={{
            _: spacer(2),
            sm: spacer(3),
            xl: spacer(4),
          }}
          gridTemplateColumns={{
            _: '100%',
            md: 'repeat(2, 1fr)',
            lg: 'repeat(3, 1fr)',
            xl: 'repeat(4, 1fr)',
          }}
          m="0 auto"
          mt={5}
          padding={0}
          width={{
            _: '100%',
            sm: '90%',
          }}
          data-testid="GitHubGrid"
          data-topic={query}
        >
          {data.map((d: Record<string, any>) => (
            <Item key={d.id} href={d.html_url} target="_blank">
              <Image src={d.owner.avatar_url} alt={d.owner.login} />
              <ItemHeader>
                <Heading as="h5" h={100} lineHeight={1}>
                  {d.name}
                </Heading>
                <small>{d.owner.login}</small>
              </ItemHeader>
              <Paragraph>{d.description}</Paragraph>
            </Item>
          ))}
        </Grid>
      );
    } else {
      output = <h3>Nothing found</h3>;
    }
  } else {
    output = <Loader block />;
  }

  return (
    <div key="GitHub" data-testid="GitHubWrapper">
      <Flex justifyContent="center">
        <ButtonGroup role="group" aria-label="GitHub Selector" data-testid="GitHubSelector">
          <Button
            busy={query === 'react' && isRunning}
            data-topic="react"
            invert={query !== 'react'}
            onClick={handleClick}
            size="lg"
          >
            React
          </Button>
          <Button
            busy={query === 'redux' && isRunning}
            data-topic="redux"
            invert={query !== 'redux'}
            onClick={handleClick}
            size="lg"
          >
            Redux
          </Button>
        </ButtonGroup>
      </Flex>
      {output}
    </div>
  );
}

export default GitHub;
