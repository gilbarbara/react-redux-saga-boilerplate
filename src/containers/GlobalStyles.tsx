import { css, Global } from '@emotion/react';

import { appColor } from '~/modules/theme';

export default function GlobalStyles() {
  return (
    <Global
      styles={css`
        *,
        *:before,
        *:after {
          box-sizing: border-box;
        }

        html {
          font-size: 62.5%;
          -webkit-font-smoothing: antialiased;
          height: 100%;
        }

        body {
          font-family: Lato, sans-serif;
          font-size: 16px; /* stylelint-disable unit-disallowed-list */
          margin: 0;
          min-height: 100vh;
          padding: 0;
        }

        img {
          height: auto;
          max-width: 100%;
        }

        a {
          color: ${appColor};
          text-decoration: none;

          &.disabled {
            pointer-events: none;
          }
        }

        button {
          appearance: none;
          background-color: transparent;
          border: 0;
          cursor: pointer;
          display: inline-block;
          font-family: inherit;
          line-height: 1;
          padding: 0;
        }
      `}
    />
  );
}
