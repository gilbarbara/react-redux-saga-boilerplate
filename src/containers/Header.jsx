import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import rgba from 'polished/lib/color/rgba';
import { appColor, headerHeight } from 'modules/theme';

import { logOut } from 'actions/index';

import { Container, utils } from 'styled-minimal';
import Icon from 'components/Icon';
import Logo from 'components/Logo';

const { responsive, spacer } = utils;

const HeaderWrapper = styled.header`
  background-color: ${rgba(appColor, 0.9)};
  height: ${headerHeight}px;
  left: 0;
  position: fixed;
  right: 0;
  top: 0;
  z-index: 200;

  &:before {
    background-color: ${appColor};
    bottom: 0;
    content: '';
    height: 0.2rem;
    left: 0;
    position: absolute;
    right: 0;
  }
`;

const HeaderLogo = styled(Logo)`
  svg {
    width: 5rem;
  }
`;

const StyledContainer = styled(Container)`
  align-items: center;
  display: flex;
  flex-wrap: wrap;
  height: 100%;
  justify-content: space-between;
  padding-bottom: ${spacer(2)};
  padding-top: ${spacer(2)};
`;

const Logout = styled.button`
  align-items: center;
  color: #333;
  display: flex;
  font-size: 1.3rem;
  padding: ${spacer(2)};

  ${responsive({ lg: 'font-size: 1.6rem;' })}; /* stylelint-disable-line */

  &.active {
    color: #000;
  }

  span {
    display: inline-block;
    margin-right: 0.4rem;
    text-transform: uppercase;
  }
`;

export default class Header extends React.PureComponent {
  static propTypes = {
    dispatch: PropTypes.func.isRequired,
  };

  handleClickLogout = () => {
    const { dispatch } = this.props;

    dispatch(logOut());
  };

  render() {
    return (
      <HeaderWrapper>
        <StyledContainer>
          <HeaderLogo />
          <Logout onClick={this.handleClickLogout}>
            <span>logout</span>
            <Icon name="sign-out" color="#333" width={16} />
          </Logout>
        </StyledContainer>
      </HeaderWrapper>
    );
  }
}
