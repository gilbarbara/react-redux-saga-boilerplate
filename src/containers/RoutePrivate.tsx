import React from 'react';
import { Redirect, Route } from 'react-router-dom';

import { RouteProps } from 'types';

function RoutePrivate({ component: Component, isAuthenticated, to = '/', ...rest }: RouteProps) {
  return (
    <Route
      {...rest}
      render={props =>
        isAuthenticated ? (
          <Component {...props} />
        ) : (
          <Redirect
            to={{
              pathname: to,
              state: { redirect: props.location.pathname, isAuthenticated },
            }}
          />
        )
      }
    />
  );
}

export default RoutePrivate;
