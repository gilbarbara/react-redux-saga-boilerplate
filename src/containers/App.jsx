import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Router, Switch, Route } from 'react-router-dom';
import Helmet from 'react-helmet';
import styled, { css, ThemeProvider } from 'styled-components';
import treeChanges from 'tree-changes';

import history from 'modules/history';
import theme, { headerHeight } from 'modules/theme';
import { utils } from 'styled-minimal';

import config from 'config';
import { showAlert } from 'actions/index';

import Home from 'routes/Home';
import Private from 'routes/Private';
import NotFound from 'routes/NotFound';

import Header from 'containers/Header';
import SystemAlerts from 'containers/SystemAlerts';

import Footer from 'components/Footer';
import GlobalStyles from 'components/GlobalStyles';
import RoutePublic from 'components/RoutePublic';
import RoutePrivate from 'components/RoutePrivate';

const StyledApp = styled.div`
  display: flex;
  flex-direction: column;
  min-height: 100vh;
  opacity: 1 !important;
  position: relative;
  transition: opacity 0.5s;
`;

const MainPrivate = ({ isAuthenticated }) =>
  isAuthenticated &&
  css`
    padding: ${utils.px(headerHeight)} 0 0;
  `;

const Main = styled.main`
  min-height: 100vh;

  ${MainPrivate};
`;

export class App extends React.Component {
  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    user: PropTypes.object.isRequired,
  };

  componentWillReceiveProps(nextProps) {
    const { dispatch } = this.props;
    const { changedTo } = treeChanges(this.props, nextProps);

    /* istanbul ignore else */
    if (changedTo('user.isAuthenticated', true)) {
      dispatch(showAlert('Hello! And welcome!', { variant: 'success', icon: 'bell' }));
    }
  }

  render() {
    const { dispatch, user } = this.props;

    return (
      <Router history={history}>
        <ThemeProvider theme={theme}>
          <StyledApp logged={user.isAuthenticated}>
            <Helmet
              defer={false}
              htmlAttributes={{ lang: 'pt-br' }}
              encodeSpecialCharacters={true}
              defaultTitle={config.title}
              titleTemplate={`%s | ${config.name}`}
              titleAttributes={{ itemprop: 'name', lang: 'pt-br' }}
            />
            {user.isAuthenticated && <Header dispatch={dispatch} user={user} />}
            <Main isAuthenticated={user.isAuthenticated}>
              <Switch>
                <RoutePublic
                  isAuthenticated={user.isAuthenticated}
                  path="/"
                  exact
                  component={Home}
                />
                <RoutePrivate
                  isAuthenticated={user.isAuthenticated}
                  path="/private"
                  component={Private}
                />
                <Route component={NotFound} />
              </Switch>
            </Main>
            <Footer />
            <SystemAlerts />
            <GlobalStyles />
          </StyledApp>
        </ThemeProvider>
      </Router>
    );
  }
}

/* istanbul ignore next */
function mapStateToProps(state) {
  return {
    user: state.user,
  };
}

export default connect(mapStateToProps)(App);
