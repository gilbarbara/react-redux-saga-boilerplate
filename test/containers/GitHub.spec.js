import React from 'react';

import { STATUS } from 'constants/index';

import { GitHub } from 'containers/GitHub';

jest.mock('uuid/v4', () => () => 'ABCDE');

const mockDispatch = jest.fn();
const props = {
  dispatch: mockDispatch,
  github: {
    repos: {
      data: [],
    },
  },
};

function setup(ownProps = props) {
  return shallow(<GitHub {...ownProps} />, { attachTo: document.getElementById('react') });
}

describe('GitHub', () => {
  const wrapper = setup();

  it('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a Loader without data', () => {
    expect(wrapper.find('Loader')).toExist();
  });

  it('should have dispatched an action on mount', () => {
    expect(mockDispatch).toHaveBeenCalledWith({
      payload: { query: 'react' },
      type: 'GITHUB_GET_REPOS',
    });
  });

  it("should not render if selected data doesn't exist", () => {
    wrapper.setProps({
      github: {
        repos: {
          data: {},
          status: STATUS.READY,
        },
      },
    });

    expect(wrapper.find('Grid')).not.toExist();
  });

  it('should render the Grid if data exists', () => {
    wrapper.setProps({
      github: {
        repos: {
          data: {
            react: [
              {
                id: 12,
                name: 'magic-tricks',
                html_url: 'https://github.com/username/magic-tricks',
                description: 'nothing much',
                owner: {
                  avatar_url: 'avatar_url',
                  login: 'username',
                },
              },
            ],
          },
          status: STATUS.READY,
        },
      },
    });

    expect(wrapper.find('Grid')).toMatchSnapshot();
  });

  it('should dispatch an alert', () => {
    wrapper.setProps({
      github: {
        repos: {
          data: {},
          status: STATUS.ERROR,
          message: 'Noothing found',
        },
      },
    });

    expect(mockDispatch).toHaveBeenCalledWith({
      type: 'SHOW_ALERT',
      payload: {
        id: 'ABCDE',
        icon: undefined,
        message: 'Noothing found',
        position: 'bottom-right',
        variant: 'danger',
        timeout: 0,
      },
    });
  });

  it('should dispatch an action when click selector button', () => {
    const button = mount(
      wrapper
        .find('ButtonGroup')
        .childAt(1)
        .getElement(),
    );

    button.simulate('click');

    expect(mockDispatch).toHaveBeenCalledWith({
      payload: { query: 'redux' },
      type: 'SWITCH_MENU',
    });
  });
});
