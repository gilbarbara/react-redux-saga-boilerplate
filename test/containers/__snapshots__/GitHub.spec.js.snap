// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitHub should render properly 1`] = `
<GitHubWrapper
  key="GitHub"
>
  <Flex
    justifyContent="center"
  >
    <ButtonGroup
      aria-label="GitHub Selector"
      role="group"
    >
      <Button
        animate={false}
        block={false}
        data-query="react"
        disabled={false}
        onClick={[Function]}
        outline={false}
        size="lg"
        type="button"
        variant="primary"
      >
        React
      </Button>
      <Button
        animate={false}
        block={false}
        data-query="redux"
        disabled={false}
        onClick={[Function]}
        outline={true}
        size="lg"
        type="button"
        variant="primary"
      >
        Redux
      </Button>
    </ButtonGroup>
  </Flex>
  <Loader
    type="pulse"
  />
</GitHubWrapper>
`;

exports[`GitHub should render the Grid if data exists 1`] = `null`;
